File: Data Structures & Algorithms/contains-duplicate-ii/submission-3.py
class Solution:
    def containsNearbyDuplicate(self, nums: List[int], k: int) -> bool:
        memo = {}
        for i, num in enumerate(nums):
            if num in memo:
                if abs(memo[num] - i) <= k:
                    return True
                else:
                    memo[num] = i
            else:
                memo[num] = i
        return False

File: Data Structures & Algorithms/house-robber/submission-3.cpp
class Solution {
public:
    vector<int> memo;
    int rob(vector<int>& nums) {
        memo.assign(nums.size(), -1);
        return helper(0, nums);
    }
    int helper(int i, vector<int> &nums) {
        if (i >= nums.size()) return 0;
        if (memo[i] >= 0) return memo[i];
        memo[i] = max(nums[i] + helper(i + 2, nums), helper(i+1, nums));
        return memo[i];
    }
};
